Philips.CodeAnalysis.Common/SingleDiagnosticAnalyzer.cs

```diff
@@ -15,14 +15,17 @@ namespace Philips.CodeAnalysis.Common
 	public abstract class SingleDiagnosticAnalyzer : DiagnosticAnalyzer
 	{
 		public DiagnosticId DiagnosticId { get; }
-
-		public DiagnosticDescriptor Rule { get; }
+		public string Id { get; }
+		protected Helper Helper { get; }
+		protected DiagnosticDescriptor Rule { get; }
 
 		protected SingleDiagnosticAnalyzer(DiagnosticId id, string title, string messageFormat, string description, string category,
-											DiagnosticSeverity severity = DiagnosticSeverity.Error, bool isEnabled = true)
+											Helper helper = null, DiagnosticSeverity severity = DiagnosticSeverity.Error, bool isEnabled = true)
 		{
 			DiagnosticId = id;
-			Rule = new(Helper.ToDiagnosticId(id), title, messageFormat, category, severity, isEnabled, description);
+			Id = Helper.ToDiagnosticId(id);
+			Rule = new(Id, title, messageFormat, category, severity, isEnabled, description);
+			Helper = helper;
```

Comment:
Place an initializer in the `Helper` helper property and remove it from the constructor parameters.

Reply:
I had this idea, but was constrained because the Rule needs to be created by the time the constructor is created, and it is immutable, and it needs Helper to get the DiagnosticId. But combined with your other idea of moving this method into the base (out of Helper), this could work. I'll try it later today.